File: item-list/src/main/java/com/gurunars/item_list/ChangeDelete.java
package com.gurunars.item_list;

import android.support.v7.widget.RecyclerView;

import java.util.List;

class ChangeDelete<ItemType extends Item> extends ChangeOfPart<ItemType> {

    ChangeDelete(ItemType item, int sourcePosition, int targetPosition) {
        super(item, sourcePosition, targetPosition);
    }

    @Override
    public int apply(RecyclerView.Adapter adapter, Scroller scroller, List<ItemType> items,
                     int currentPosition) {
        items.remove(sourcePosition);
        adapter.notifyItemRemoved(sourcePosition);
        return currentPosition;
    }
}

File: item-list/src/main/java/com/gurunars/item_list/ChangeDelete.kt
package com.gurunars.item_list

import android.support.v7.widget.RecyclerView

internal class ChangeDelete<ItemType : Item>(item: ItemType, sourcePosition: Int, targetPosition: Int) : ChangeOfPart<ItemType>(item, sourcePosition, targetPosition) {

    override fun apply(adapter: RecyclerView.Adapter<*>, scroller: Scroller, items: MutableList<ItemType>,
                       currentPosition: Int): Int {
        items.removeAt(sourcePosition)
        adapter.notifyItemRemoved(sourcePosition)
        return currentPosition
    }
}
